src/django_components/slots.py

```diff
@@ -0,0 +1,467 @@
+            Hello!
+        {% endfill %}
+        {% fill "second_fill" %}
+            Hello too!
+        {% endfill %}
+    {% endcomponent %}
+    ```
+    Then this function returns the nodes (`django.template.Node`) for `fill "first_fill"`
+    and `fill "second_fill"`.
+    """
+    fill_nodes: Union[Iterable[NamedFillNode], ImplicitFillNode] = []
+    if _block_has_content(component_nodelist):
+        for parse_fn in (
+            _try_parse_as_default_fill,
+            _try_parse_as_named_fill_tag_set,
+        ):
+            curr_fill_nodes = parse_fn(component_nodelist, ComponentNodeCls)
+            if curr_fill_nodes:
+                fill_nodes = curr_fill_nodes
+                break
+        else:
+            raise TemplateSyntaxError(
+                "Illegal content passed to 'component' tag pair. "
+                "Possible causes: 1) Explicit 'fill' tags cannot occur alongside other "
+                "tags except comment tags; 2) Default (default slot-targeting) content "
+                "is mixed with explict 'fill' tags."
+            )
+    return fill_nodes
+
+
+def _try_parse_as_named_fill_tag_set(
+    nodelist: NodeList,
+    ComponentNodeCls: Type[Node],
+) -> Optional[Iterable[NamedFillNode]]:
+    result = []
+    seen_name_fexps: Set[FilterExpression] = set()
+    for node in nodelist:
+        if isinstance(node, NamedFillNode):
+            if node.name_fexp in seen_name_fexps:
+                raise TemplateSyntaxError(
+                    f"Multiple fill tags cannot target the same slot name: "
+                    f"Detected duplicate fill tag name '{node.name_fexp}'."
+                )
+            seen_name_fexps.add(node.name_fexp)
+            result.append(node)
+        elif isinstance(node, CommentNode):
+            pass
+        elif isinstance(node, TextNode) and node.s.isspace():
+            pass
+        else:
+            return None
+    return result
+
+
+def _try_parse_as_default_fill(
+    nodelist: NodeList,
+    ComponentNodeCls: Type[Node],
+) -> Optional[ImplicitFillNode]:
+    nodes_stack: List[Node] = list(nodelist)
+    while nodes_stack:
+        node = nodes_stack.pop()
+        if isinstance(node, NamedFillNode):
+            return None
+        elif isinstance(node, ComponentNodeCls):
+            # Stop searching here, as fill tags are permitted inside component blocks
+            # embedded within a default fill node.
+            continue
+        for nodelist_attr_name in node.child_nodelists:
+            nodes_stack.extend(getattr(node, nodelist_attr_name, []))
+    else:
+        return ImplicitFillNode(nodelist=nodelist)
+
+
+def _block_has_content(nodelist: NodeList) -> bool:
+    for node in nodelist:
+        if isinstance(node, TextNode) and node.s.isspace():
+            pass
+        elif isinstance(node, CommentNode):
+            pass
+        else:
+            return True
+    return False
+
+
+def render_component_template_with_slots(
```

Comment:
NOTE: I've moved the last chunk of `Component.render` logic here into `render_component_template_with_slots`. The reason being that before, the flow was a bit unclear, because the template instance as owned/attached by `Component` class, but then was modified by the slot fills, and then `Component` rendered it. So there was hopping back and forth.

Now with `render_component_template_with_slots` here, `Component` just calls it and expects to get back a string. And so `Component` doesn't need to know how to slots logic is implemented.